gluonnlp/model/block.py

```diff
@@ -89,3 +90,100 @@ def __init__(self, axis=-1, eps=1E-6, **kwargs):
     def hybrid_forward(self, F, x):  # pylint: disable=arguments-differ
         ret = F.broadcast_div(x, F.norm(x, axis=self._axis, keepdims=True) + self._eps)
         return ret
+
+
+class CharacterLevelCNNEmbedding(HybridBlock):
+    """
+    CharacterLevelCNNEmbedding takes data and embed it using 2D convolution layers. Quantity of
+    Conv2D block is equal to len of channels and kernel_size lists, which should be same.
+
+    Each item in the data is an index of a character, drawn from a vocab of `vocab_size`.
+    Input layout: NTC, where
+    N - batch_size = # of examples,
+    T - seq_len    = # of words in each example,
+    C - channel    = # char in each word
+
+    Output layout: batch_size x seq_len x channels, where
+    batch_size = # of examples,
+    seq_len    = # words in each example,
+    channels   = # channels equal to `channels` parameter passed to `__init__()`
+    """
+    def __init__(self, channels, kernel_sizes, padding, vocab_size, char_embedding_size=8,
+                 keep_prob=0.2, prefix=None, params=None):
+        super(CharacterLevelCNNEmbedding, self).__init__()
+
+        assert len(channels) == len(kernel_sizes)
+        self.net = nn.HybridSequential(prefix=prefix, params=params)
+        self.embedding = nn.Embedding(input_dim=vocab_size, output_dim=char_embedding_size)
+
+        with self.net.name_scope():
+            for channel, kernel in zip(channels, kernel_sizes):
+                self.net.add(nn.Dropout(rate=keep_prob))
+                self.net.add(nn.Conv2D(in_channels=char_embedding_size,
+                                       strides=[1, 1], channels=channel, kernel_size=[1, kernel],
+                                       padding=padding))
+
+    def hybrid_forward(self, F, x):  # pylint: disable=arguments-differ
+        x_embedded = self.embedding(x)
+        # transposing to NCHW which is a default for Conv2D class
+        x_embedded_transposed = x_embedded.transpose(axes=(0, 3, 1, 2))
+        network_output = self.net.hybrid_forward(F, x_embedded_transposed)
+        relu_output = F.relu(network_output)
+        max_output = F.max(relu_output, axis=3)
+        # transposing back to expected batch_size x seq_len x channels
+        return max_output.transpose(axes=(0, 2, 1))
+
+
+class PredefinedEmbedding(HybridBlock):
```

Comment:
This embedding is not necessary since we can directly initialize a word embedding layer using the glove vectors: `net.embedding.weight.set_data(glove_embedding.idx_to_vec.as_in_context(ctx))`